File: cpp/anagram/anagram.cpp
#include <iostream>
#include <fstream>
#include <set>

using namespace std;

// Reads all known words from a dictionary file
// (based on the Ubuntu wamerican package) and
// returns the set containing those words.
set<string> getWords () {
	ifstream inputStream("dictionary.txt");
	set<string> wordSet;

	string word;
	while (inputStream >> word) {
		wordSet.insert(word);
	}

	return wordSet;
}

// Collects user input and enforces
// lower case for later comparisons.
string getInput () {
	string input;
	cin >> input;

	// Make user input lowercase for
	// dictionary comparisons.
	for (char &c : input) {
		c = tolower(c);
	}

	return input;
}

// Generates and adds all permutations of the given word to given set.
// 
// HT: https://www.techiedelight.com/find-permutations-given-string/
void getPermutations(string word, int head, int tail, set<string>* perm) {
	if (head == tail - 1) {
		(*perm).insert(word);
		return;
	}

	for (int current = head; current < tail; current++) {
		swap(word[head], word[current]);
		getPermutations(word, head + 1, tail, perm);
	}
}

int main (int argc, char *argv[]) {
	cout << "Loading dictionary, please wait...\n";
	set<string> wordSet = getWords();

	cout << "Enter a word: ";
	string baseWord = getInput();

	set<string> permutations;
	getPermutations(baseWord, 0, baseWord.length(), &permutations);

	// Assume there are no matches
	// until proven otherwise.
	bool hasAnagram = false;

	cout << "Anagrams: ";
	for ( string wordPerm : permutations ) {
		// Note: a word is allowed to be its own anagram
		// as long as it appears in the dictionary. This
		// is to distinguish the baseWord from fake works.
		if (wordSet.find(wordPerm) != wordSet.end()) {
			// Display valid anagrams.
			cout << wordPerm << " ";
			hasAnagram = true;
		}
	}
	cout << ( hasAnagram ? "\n" : "NONE\n" );

	return EXIT_SUCCESS;
}

File: python/anagram/anagram.py
#!/usr/bin/env python3
import itertools


# Returns a set of valid words from a dictionary file,
# which is based on the Ubuntu wamerican package.
def load_dictionary():
    print("Loading dictionary, please wait...")
    f = open("dictionary.txt")
    dictionary = set([word.strip() for word in f.readlines()])
    f.close()
    return dictionary


# Reads user input as a lowercase string.
def get_word():
    word = ""
    while word == "":
        word = input("Enter a word: ")
    return word.lower()


# Given a word and a dictionary (as a set of words),
# this function returns all permutations of the word
# that are valid within the given dictionary.
def get_anagrams(word, dictionary):
    anagrams = set([])
    permutations = set(["".join(perm) for perm in itertools.permutations(word)])
    for perm in permutations:
        if perm != word and perm in dictionary:
            anagrams.add(perm)
    return anagrams


def main():
    dictionary = load_dictionary()
    word = get_word()
    anagrams = get_anagrams(word, dictionary)
    print("Anagrams: %s" % (", ".join(anagrams) or "none"))


if __name__ == "__main__":
    main()
